pkg/filewatcher/filewatcher.go

```diff
@@ -0,0 +1,75 @@
+package filewatcher
+
+import (
+	"fmt"
+	"hash/fnv"
+	"os"
+	"sync"
+	"time"
+
+	"k8s.io/klog/v2"
+)
+
+var (
+	initialFileHash          string
+	watchCertificateFileOnce sync.Once
+)
+
+func WatchFileForChanges(fileToWatch string, interval time.Duration) error {
+	var fileContents []byte
+	var err error
+
+	watchCertificateFileOnce.Do(func() {
+		klog.Infof("Starting the file change watcher on file, %s", fileToWatch)
+		fileContents, err = os.ReadFile(fileToWatch)
+		if err != nil {
+			err = fmt.Errorf("failed to read file on initialization: %v", err)
+			return
+		}
+
+		initialFileHash = hashSimple(fileContents)
+		go checkForFileChanges(fileToWatch, interval)
+	})
+	return err
+}
+
+func checkForFileChanges(fileToWatch string, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	done := make(chan bool)
+
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				klog.Infof("Checking file for changes, %s", fileToWatch)
+				fileContents, err := os.ReadFile(fileToWatch)
+				if err != nil {
+					klog.Error("failed to read the file: %v", err)
```

Comment:
When a file read fails we will just stop watching it and if it changes this will never be caught.

Reply:
The return should just return from the sync.Once function and the error would be returned at the end of `WatchFileForChanges`